ParallelPrimitives/RadixSortKernels.h

```diff
@@ -521,278 +514,708 @@ extern "C" __global__ void SortSinglePassKernel( int* gSrcKey, int* gDstKey, int
-{
-	if( threadId == 0 )
+	__syncthreads();
+
+	uint32_t numberOfBlocks = div_round_up( numberOfInputs, GHISTOGRAM_ITEM_PER_BLOCK );
+	__shared__ uint32_t iBlock;
+	if( threadIdx.x == 0 )
+	{
+		iBlock = atomicInc( counter, 0xFFFFFFFF );
+	}
+
+	__syncthreads();
+
+	bool hasData = false;
+
+	while( iBlock < numberOfBlocks )
 	{
-		int offset = 0;
+		hasData = true;
 
-		if( blockId != 0 )
+#if defined( KEY_IS_16BYTE_ALIGNED )
+		if( ( iBlock + 1 ) * GHISTOGRAM_ITEM_PER_BLOCK <= numberOfInputs )
 		{
-			while( !gIsReady[blockId - 1] )
+			for( int i = 0; i < GHISTOGRAM_ITEM_PER_BLOCK; i += GHISTOGRAM_THREADS_PER_BLOCK * 4 )
 			{
+				uint32_t itemIndex = iBlock * GHISTOGRAM_ITEM_PER_BLOCK + i + threadIdx.x * 4;
+				struct alignas( 16 ) Key4
+				{
+					RADIX_SORT_KEY_TYPE xs[4];
+				};
+				Key4 key4 = *(Key4*)&inputs[itemIndex];
+				for( int k = 0; k < 4; k++ )
+				{
+					auto item = key4.xs[k];
+					for( int i = 0; i < sizeof( RADIX_SORT_KEY_TYPE ); i++ )
+					{
+						uint32_t bitLocation = startBits + i * 8;
+						uint32_t bits = ( getKeyBits( item ) >> bitLocation ) & 0xFF;
+						atomicInc( &localCounters[i][bits], 0xFFFFFFFF );
+					}
+				}
+			}
+		}
+		else
+#endif
+			for( int i = 0; i < GHISTOGRAM_ITEM_PER_BLOCK; i += GHISTOGRAM_THREADS_PER_BLOCK )
+			{
+				uint32_t itemIndex = iBlock * GHISTOGRAM_ITEM_PER_BLOCK + threadIdx.x + i;
+				if( itemIndex < numberOfInputs )
+				{
+					auto item = inputs[itemIndex];
+					for( int i = 0; i < sizeof( RADIX_SORT_KEY_TYPE ); i++ )
+					{
+						uint32_t bitLocation = startBits + i * 8;
+						uint32_t bits = ( getKeyBits( item ) >> bitLocation ) & 0xFF;
+						atomicInc( &localCounters[i][bits], 0xFFFFFFFF );
+					}
+				}
 			}
 
-			offset = gPartialSum[blockId - 1];
-
-			__threadfence();
+		__syncthreads();
 
-			// Reset the value
-			gIsReady[blockId - 1] = false;
+		if( threadIdx.x == 0 )
+		{
+			iBlock = atomicInc( counter, 0xFFFFFFFF );
 		}
 
-		gPartialSum[blockId] = offset + currentSegmentSum;
-
-		// Ensure that the gIsReady is only modified after the gPartialSum is written.
-		__threadfence();
+		__syncthreads();
+	}
 
-		gIsReady[blockId] = true;
+	if( hasData )
+	{
+		__syncthreads();
 
-		*currentGlobalOffset = offset;
+		for( int i = 0; i < sizeof( RADIX_SORT_KEY_TYPE ); i++ )
+		{
+			for( int j = threadIdx.x; j < 256; j += GHISTOGRAM_THREADS_PER_BLOCK )
+			{
+				atomicAdd( &gpSumBuffer[256 * i + j], localCounters[i][j] );
+			}
+		}
 	}
-
-	__syncthreads();
 }
 
-extern "C" __global__ void ParallelExclusiveScanAllWG( int* gCount, int* gHistogram, volatile int* gPartialSum, volatile bool* gIsReady )
+extern "C" __global__ void gPrefixSum( uint32_t* gpSumBuffer )
 {
-	// Fill the LDS with the partial sum of each segment
-	__shared__ int blockBuffer[SCAN_WG_SIZE];
+	__shared__ uint32_t smem[256];
```

Comment:
Where is this 256 from ?  Does it have a variable / constant associated to it ?  Please used that named variable instead.

Reply:
replaced by BIN_SIZE